src/python/WMQuality/Emulators/RucioClient/MockRucioApi.py

```diff
@@ -0,0 +1,155 @@
+        return genericLookup
+
+    def getDataLockedAndAvailable(self, **kwargs):
+        """
+        Mock the method to discover where data is locked and available.
+        Note that, by default, it will not return any Tape RSEs.
+        :return: a unique list of RSEs
+        """
+        if 'name' not in kwargs:
+            raise WMRucioException("A DID name must be provided to the getBlockLockedAndAvailable API")
+        if self.isContainer(kwargs['name']):
+            # then resolve it at container level and all its blocks
+            return self.getContainerLockedAndAvailable(**kwargs)
+
+        if 'grouping' in kwargs:
+            # long strings seem not to be working, like ALL / DATASET. Make it short!
+            kwargs['grouping'] = kwargs['grouping'][0]
+        # TODO: either grouping or returnTape should change this response...
+        returnTape = kwargs.pop("returnTape", False)
+
+        rses = set()
+        if kwargs['name'].split('#')[0] == PILEUP_DATASET:
+            # Pileup is at a single site
+            sites = ['T2_XX_SiteC']
+            _BLOCK_LOCATIONS[kwargs['name']] = sites
+        else:
+            sites = self.sitesByBlock(block=kwargs['name'])
+            _BLOCK_LOCATIONS[kwargs['name']] = sites
+        rses.update(sites)
+        return list(rses)
+
+
+    def getContainerLockedAndAvailable(self, **kwargs):
+        """
+        Mock the method to discover where container data is locked and available.
+        Note that, by default, it will not return any Tape RSEs.
+        :return: a unique list of RSEs
+        """
+        if 'name' not in kwargs:
+            raise WMRucioException("A DID name must be provided to the getContainerLockedAndAvailable API")
+        kwargs.setdefault("scope", "cms")
+        if 'grouping' in kwargs:
+            # long strings seem not to be working, like ALL / DATASET. Make it short!
+            kwargs['grouping'] = kwargs['grouping'][0]
+        # TODO: either grouping or returnTape should change this response...
+        returnTape = kwargs.pop("returnTape", False)
+
+        if kwargs['name'] == PILEUP_DATASET:
+            return ['T2_XX_SiteA', 'T2_XX_SiteB', 'T2_XX_SiteC']
+        try:
+            DBS3Reader(PROD_DBS).checkDatasetPath(kwargs['name'])
+            blocks = DBS3Reader(PROD_DBS).dbs.listBlocks(dataset=kwargs['name'])
+            singleBlock = blocks[0]['block_name']
+            return self.sitesByBlock(singleBlock)
+        except DBSReaderError:
+            return []
+
+    def isContainer(self, didName, scope='cms'):
+        """
+        Checks whether the DID name corresponds to a container type or not.
+        :param didName: string with the DID name
+        :param scope: string containing the Rucio scope (defaults to 'cms')
+        :return: True if the DID is a container, else False
+        """
+        # TODO: figure use cases where we need to raise this exception
+        if didName == "a bad DID name yet to be defined":
```

Comment:
Do we not have a proper regular Expression describing a correct DID name?

Reply:
Yes, but the problem is that this is supposed to catch a DID that does not exist in Rucio. That means we would have to validate the input against Container, Block and LFN, before we can raise the exception.

This still needs to be further implemented, right now we only evaluate whether it's a block or a container (in the old and dirty way!).

Reply:
Agreed on that too.